taichi/ir/control_flow_graph.cpp

```diff
@@ -863,4 +864,133 @@ std::unordered_set<SNode *> ControlFlowGraph::gather_loaded_snodes() {
+  // pre-processing step for the next maximum stack size determining algorithm.
+  for (int i = 0; i < num_nodes; i++) {
+    for (int j = nodes[i]->begin_location; j < nodes[i]->end_location; j++) {
+      Stmt *stmt = nodes[i]->block->statements[j].get();
+      if (auto *stack_push = stmt->cast<AdStackPushStmt>()) {
+        auto *stack = stack_push->stack->as<AdStackAllocaStmt>();
+        if (stack->max_size == 0 /*adaptive*/) {
+          increased_size[i][stack]++;
+          if (increased_size[i][stack] > max_increased_size[i][stack]) {
+            max_increased_size[i][stack] = increased_size[i][stack];
+          }
+        }
+      } else if (auto *stack_pop = stmt->cast<AdStackPopStmt>()) {
+        auto *stack = stack_pop->stack->as<AdStackAllocaStmt>();
+        if (stack->max_size == 0 /*adaptive*/) {
+          increased_size[i][stack]--;
+        }
+      }
+    }
+    to_visit.push(i);
+    in_queue[i] = true;
+  }
+
+  // The maximum stack size determining algorithm -- an algorithm similar to
+  // the Bellman-Ford algorithm.
+  while (!to_visit.empty()) {
+    int node_id = to_visit.front();
+    to_visit.pop();
+    in_queue[node_id] = false;
+    CFGNode *now = nodes[node_id].get();
+
+    for (auto &it : max_increased_size[node_id]) {
+      // Inside this CFGNode -- update the answer |max_size|
+      auto *stack = it.first;
+      auto max_size_inside_this_node = it.second;
+      auto current_max_size =
+          max_size_at_node_begin[node_id][stack] + max_size_inside_this_node;
+      if (current_max_size > max_ad_stack_size) {
+        current_max_size = max_ad_stack_size;
+        oversized_stacks.insert(stack);
+      }
+      if (current_max_size > max_size[stack]) {
+        max_size[stack] = current_max_size;
+      }
+    }
+    for (auto &it : increased_size[node_id]) {
+      // At the end of this CFGNode -- update the state |max_size_at_node_begin|
+      // of other CFGNodes
+      auto *stack = it.first;
+      auto increase_in_this_node = it.second;
+      auto current_size =
+          max_size_at_node_begin[node_id][stack] + increase_in_this_node;
+      if (current_size > max_ad_stack_size) {
+        current_size = max_ad_stack_size;  // avoid infinite loop
+      }
+      for (auto *next_node : now->next) {
+        int next_node_id = node_ids[next_node];
+        if (current_size > max_size_at_node_begin[next_node_id][stack]) {
+          max_size_at_node_begin[next_node_id][stack] = current_size;
+          if (!in_queue[next_node_id]) {
+            to_visit.push(next_node_id);
+            in_queue[next_node_id] = true;
+          }
+        }
+      }
+    }
+  }
+
+  if (!oversized_stacks.empty()) {
+    std::vector<std::string> oversized_stacks_name;
+    oversized_stacks_name.reserve(oversized_stacks.size());
+    for (auto &stack : oversized_stacks) {
+      oversized_stacks_name.push_back(stack->name());
+    }
+    TI_WARN(
```

Comment:
How serious it is if the necessary AD stack size overflows `max_ad_stack_size`? If this would directly result in wrong results, we should better report an error and stop immediately?

Reply:
This is not that serious -- IMHO in most cases, it's the control-flow graph that unable to determine the max capacity. For example:
```
s = stack()
for i in range(10):
  s.push(i)
```
The control-flow graph does not have the range `10`, and thus cannot determine the capacity.

From another perspective, the current codebase uses a fixed capacity of 16, and it works fine.

On the other hand, I don't see any warnings in the current tests, so maybe even the above case (a loop with #pushes > #pops) doesn't exist, and the control-flow graph is able to determine all maximum capacities. Whether the above case exists depends on the usage of AD-stack in auto_diff.cpp.

Reply:
Hmm, the code implementation looks like it is able to figure out the necessary stack size, which overflows `max_ad_stack_size`. But the interpretation is "unable to determine the max capacity". Is it possible to distinguish these two cases, or maybe I'm misunderstanding something ? :rofl:

Reply:
Oh I was using the terms "size" and "capacity" interchangeably... It should be the necessary stack size, not the max capacity.

Reply:
If auto_diff.cpp is bug-free, then #pushes should be always equal to #pops?

I still don't see why the fact that an undetermined stack would not lead to a bad result. IIUC, this is serious, but there are cases where this stack size just cannot be determined statically? (Like the `range(10)` example you gave).

> I don't see any warnings in the current tests

Were you referring to the python tests, or the new CPP tests? If the former, i think that's because the output are not shown in `pytest` by default..

Reply:
> If auto_diff.cpp is bug-free, then #pushes should be always equal to #pops?
>
> I still don't see why the fact that an undetermined stack would not lead to a bad result. IIUC, this is serious, but there are cases where this stack size just cannot be determined statically? (Like the range(10) example you gave).

Right, #pushes should be always equal to #pops in auto_diff.cpp, and #pushes can be larger than #pops because unused pops at the end (i.e., the last operation to a stack is a pop) are eliminated for optimization.
There are cases where this stack size just cannot be determined statically if auto_diff.cpp can generate code like the range(10) example.

> If the former, i think that's because the output are not shown in pytest by default..

The former... How to show the outputs in these tests? `-v`?

Reply:
https://stackoverflow.com/questions/14405063/how-can-i-see-normal-print-output-created-during-pytest-run ? (You may have to tweak `ti` to make this work :rofl: )

Reply:
OK I see, but I thought `oversized_stacks` means a different thing? I.e., "I am able to figure out the stack size statically, and it definitely overflows the configured max". So capping it to `max_ad_stack_size` will lead to wrong result?

Reply:
No, `oversized_stacks` doesn't mean that "I am able to figure out the stack size statically". There are two cases in it, one is "I am able to figure out the stack size statically, and it definitely overflows the configured max", and the other is "I am unable to figure out the stack size statically". It is hard to distinguish these two cases, and I think the former cases seldom happens when `max_ad_stack_size` is relatively large (larger than actually needed).

There is another way to distinguish these cases, but the time complexity will be O(n^2) where n is the number of CFGNodes (basic blocks): if a node is pushed into the queue `to_visit` more than n times, then we have a positive cycle in the graph (the Bellman-Ford algorithm!), so we are "unable to figure out the stack size statically". Otherwise, we are "able to figure out the stack size statically", and there's no need of the compile config `max_ad_stack_size`.

Reply:
Thanks for the explanation. Let's add this comment to the code?

Reply:
Now I think maybe it's better to use the Bellman-Ford algorithm -- when we are "able to figure out **all** stack sizes statically", we are guaranteed to figure them out even if a stack needs a large size; when we are unable to figure out at least one stack size, the algorithm will run slower but we will be sure that we are unable to figure out the stack size statically (instead of a warning about possible overflow). WDYT?

Reply:
> Now I think maybe it's better to use the Bellman-Ford algorithm 

Sounds great!

> So, in this case, I think the message should not be a TI_WARN -- maybe a TI_INFO or TI_DEBUG. WDYT?

IIUC, we cannot determine the stack size as soon as the kernel has a loop in it? If so, yeah, maybe it's better to make this TI_DEBUG to reduce the noise..

Reply:
> IIUC, we cannot determine the stack size as soon as the kernel has a loop in it? If so, yeah, maybe it's better to make this TI_DEBUG to reduce the noise..

Right...

Reply:
@ljcc0930  will help review Bellman-Ford XD

Reply:
LGTM!